### PR TITLE
Replace all locale-dependent operations with locale-agnostic counterparts

### LINKED ISSUE #12720: Use invalid locale instead of "C" to detect and fix any attempted locale-dependent operations in the codebase
Requested in https://github.com/ethereum/solidity/pull/12716#issuecomment-1054141878.
Related to #12350.

> I would actually be happier if we could find a way to somehow set an "illegal" locale so that we get an error whenever we perform a locale-dependent operation (and of course remove all locale-dependent operations).

### PR DESCRIPTION
**Pull request by wechman**
fixes #12720

### CODE REVIEW

**Review by chriseth — COMMENTED** [review]
I think it would be nice to have functions `toUpper(char)` and `toUpper(string)` (and lower) in our string utils header or something - same for isDigit.

File: libyul/backends/evm/EVMDialect.cpp [hunk 1]
```diff
@@ -142,8 +142,8 @@ map<YulString, BuiltinFunctionForEVM> createBuiltins(langutil::EVMVersion _evmVe
 	map<YulString, BuiltinFunctionForEVM> builtins;
 	for (auto const& instr: evmasm::c_instructions)
 	{
-		string name = instr.first;
-		transform(name.begin(), name.end(), name.begin(), [](unsigned char _c) { return tolower(_c); });
+		string name = instr.first;
```
Comment (stackenbotten) [inline libyul/backends/evm/EVMDialect.cpp]: Coding style error

**Comment by wechman** [comment]
[@chriseth](https://github.com/chriseth) I created string util functions as you suggested. However, [CommonSyntaxTest.cpp](https://github.com/ethereum/solidity/blob/develop/test/CommonSyntaxTest.cpp) still needs small cleanup around `isDigit` function. It is related to `using namespace` directive abuse I have found. The problem is described in #12743.

File: libsolutil/StringUtils.h [hunk 2]
```diff
@@ -197,4 +199,45 @@ inline std::optional<unsigned> toUnsignedInt(std::string const& _value)
...
+/// @param _c value to be converted
+/// @return the converted value
+inline char toUpper(char _c)
+{
+	return toupper(_c, std::locale::classic());
+}
+
+/// Converts parameter _s to its lowercase equivalent. It uses the classic "C" locale semantics.
+/// @param _s value to be converted
+inline void toLower(std::string& _s)
```
Comment (chriseth) [inline libsolutil/StringUtils.h]: Can you change this so that it returns a copy of the string? This way it might be rather error prone.

Comment (wechman) [reply libsolutil/StringUtils.h]: Sure! Fixup commit already added to the PR.

File: libsolutil/StringUtils.h [hunk 3]
```diff
@@ -197,4 +199,47 @@ inline std::optional<unsigned> toUnsignedInt(std::string const& _value)
...
+}
+
+/// Converts parameter _s to its lowercase equivalent. It uses the classic "C" locale semantics.
+/// @param _s value to be converted
+/// @return the converted value
+inline std::string toLower(std::string _s)
+{
+	std::transform(_s.begin(), _s.end(), _s.begin(), [](char _c) {
+		return toLower(_c);
+	});
```
Comment (ekpyron) [inline libsolutil/StringUtils.h:225]: Doesn't this work?
```suggestion
	std::transform(_s.begin(), _s.end(), _s.begin(), toLower);
```

Comment (ekpyron) [reply libsolutil/StringUtils.h:225]: Ah, it doesn't because it's overloaded... yeah, then the current way is probably nicest. So sorry, nevermind :-).

**Comment by chriseth** [comment]
Can you squash this please?

**Review by chriseth — APPROVED** [review]

**Pull request merged**